File: easy/543. Diameter of Binary Tree/543. Diameter of Binary Tree.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = 0

        def run(tree: Optional[TreeNode]):
            nonlocal diameter
            if not tree:
                return 0
            if not tree.left and not tree.right:
                return 1
            
            left = run(tree.left)
            right = run(tree.right)
            
            max_depth = max(left, right)
            diameter = max(diameter, left + right)
            return max_depth

        run(root)

        return diameter

File: easy/543. Diameter of Binary Tree/543. Diameter of Binary Tree.ts
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function diameterOfBinaryTree(root: TreeNode | null): number {
    let diameter = 0

    const run = (tree: TreeNode | null): number => {
        if (!tree)
            return 0
        if (!tree.left && !tree.right)
            return 1

        const left = run(tree.left)
        const right = run(tree.right)

        const max_depth = Math.max(left, right) + 1
        diameter = Math.max(diameter, left + right)
        return max_depth
    }

    run(root)

    return diameter
};

export {}